### DISCUSSION TITLE
Supporting LIRS replacement algorithm in CacheLib

### DISCUSSION THREAD

**Discussion by xingshengzhao** [facebook/CacheLib#99] [Ideas]
We has implemented the [LIRS replacement](https://en.wikipedia.org/wiki/LIRS_caching_algorithm) in the cachelib source code and use `cachebench` to read the traces and generate the results.

The traces in the test are from [CloudPhysics](https://www.usenix.org/conference/atc17/technical-sessions/presentation/waldspurger) and were collected in caching analytics service in production VMware environments (As the trace files only contain block number, we set `enableLookaside` to `true`, and the value size is fixed at 1KB). Some trace files can be downloaded at [here](https://drive.google.com/drive/folders/1vvnhze4UqCDk4ZnwHTPL6zarA2d9YHaR?usp=sharing)

Attached are LIRS's miss ratio curves in comparison with replacement algorithms currently in the cachelib, including LRU, LRU2Q, and TinyLFU (All algorithms simply adopt their default settings). 

The results are very encouraging: where cachelib's existing algorithms don't perform well as expected (miss ratio curve don't decrease with increase of cache size), LIRS mostly fixes the performance issue. Some of the results are attached. Please take a look.
![cachelib_results](https://user-images.githubusercontent.com/8392422/146061565-864ac8a3-4c30-4dba-99ef-99bb118c35e3.png)

[cachelib_results.pdf](https://github.com/facebook/CacheLib/files/7714036/cachelib_results.pdf)

The LIRS's source code for cachelib can be found at https://github.com/hansonzhao007/CacheLib, (in branch 'LIRS". Current implementation is a prototype, so it may not pass the unit-test in multithreading tests.)

To reproduce the results, you may follow these steps:
```
git clone --recurse-submodules https://github.com/hansonzhao007/CacheLib.git
cd CacheLib
git checkout LIRS
# external lib branch when testing
# fbthrift: 0b008a959
# fizz: b5cd3b59
# fmt: 094b66e8
# folly: c4d6fcdde
# gflags: e171aa2
# glog: 8f9ccfe
# googletest: 1b18723e
# sparsemap: 1186290
# wangle: 9f706148
# zstd: a488ba11
./contrib/build.sh -j
cp cachelib/cachebench/config.json ./build-cachelib/cachebench
cp cachelib/cachebench/run.py ./build-cachelib/cachebench
cd build-cachelib/cachebench
# before running the script, please modify the `trace_path` parameter in `run.py`
python3 run.py
```

We intend to contribute to the cachelib community by hopefully merging lirs into cachelib. To this end, we may continue to improve the lirs implementation by hearing from you and the rest of the community.

**Comment by sathyaphoenix** [comment]
@hansonzhao007  Thanks for taking the effort to implement this in CacheLib and doing some benchmarking. First, I'd like to move this to the discussion channel since the issue forum is to track immediate actionable things and we might want to discuss more on the implementation details before we merge this PR.

**Reply by xingshengzhao** [reply]
Thanks. In the next step, I will focus on limiting the memory footprint of the metadata in LIRS and support multithreading tests.

**Reply by therealgymmy** [reply]
@hansonzhao007: Hi, did you end up making these improvements to LIRS?

**Reply by xingshengzhao** [reply]
> @hansonzhao007: Hi, did you end up making these improvements to LIRS?

Hi, thanks for reaching out. I haven't had time to implement the metadata optimization for LIRS. Another member in our lab (Dr. Song Jiang, UTA) may pick up the remaining works after my graduation.

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
